src/engine/client/audio.cpp

```diff
@@ -0,0 +1,330 @@
+
+        al_extensions_[AUDIO_EXT_EFX] = true;
+    }
+
+    int attr[]={
+            ALC_FREQUENCY, 44100, /* todo: tune this */
+            ALC_MAX_AUXILIARY_SENDS, 4,
+            0
+    };
+
+    ALCcontext *context = alcCreateContext(device, attr);
+    if(context == nullptr || !alcMakeContextCurrent(context)) {
+        Error("failed to create audio context, aborting audio initialisation!\n");
+    }
+
+    if(alIsExtensionPresent("AL_SOFT_buffer_samples")) {
+        LogInfo("AL_SOFT_buffer_samples detected\n");
+        al_extensions_[AUDIO_EXT_SOFT_BUFFER_SAMPLES] = true;
+    }
+
+    alDopplerFactor(4.f);
+    alDopplerVelocity(350.f);
+}
+
+AudioManager::~AudioManager() {
+    LogInfo("shutting down audio sub-system...\n");
+
+    //Audio_StopSounds();
+    //Audio_ClearSamples();
+
+    ALCcontext *context = alcGetCurrentContext();
+    if(context != nullptr) {
+        alcMakeContextCurrent(nullptr);
+        alcDestroyContext(context);
+
+        ALCdevice *device = alcGetContextsDevice(context);
+        if(device != nullptr) {
+            alcCloseDevice(device);
+        }
+    }
+}
+
+void AudioManager::CacheSample(const std::string &path, bool preserve) {
+    if(plIsEmptyString(path)) {
+        LogWarn("invalid path, aborting!\n");
+        return;
+    }
+
+    auto i = samples_.find(path);
+    if(i != samples_.end()) {
+        LogWarn("attempted to double cache audio sample, \"%s\"!\n", path.c_str());
+        return;
+    }
+
+    /* todo: ensure the path is under <basedir/campaign>/audio/ ? */
+
+    SDL_AudioSpec spec;
+    uint32_t length;
+    uint8_t *buffer;
+    if(SDL_LoadWAV(u_find(path.c_str()), &spec, &buffer, &length) == nullptr) {
+        LogWarn("failed to load \"%s\"!\n", path);
+        return;
+    }
+
+    /* translate the spec over to oal
+     * todo: conversion... https://github.com/solemnwarning/armageddon-recorder/blob/master/src/resample.hpp#L42
+     * */
+    unsigned int format = 0;
+    switch(spec.format) {
+        case AUDIO_U8:
+            if(spec.channels == 1) {
+                format = AL_FORMAT_MONO8;
+            } else if(spec.channels == 2) {
+                format = AL_FORMAT_STEREO8;
+            }
+            break;
+        case AUDIO_S16:
+            if(spec.channels == 1) {
+                format = AL_FORMAT_MONO16;
+            } else if(spec.channels == 2) {
+                format = AL_FORMAT_STEREO16;
+            }
+            break;
+        default:break;
+    }
+
+    if(format == 0) {
+        LogWarn("invalid audio format for \"%s\"!\n", path);
+        SDL_FreeWAV(buffer);
+        return;
+    }
+
+    unsigned int buf_index;
+    alGenBuffers(1, &buf_index);
+    OALCheckErrors();
+    alBufferData(buf_index, format, buffer, length, spec.freq);
+    OALCheckErrors();
+
+    AudioSample sample;
+    sample.data = buffer;
+    sample.al_buffer_id = buf_index;
+    sample.preserve = preserve;
+    samples_.insert(std::make_pair(path, sample));
```

Comment:
~AudioSample() should probably destroy the OpenAL and SDL buffers.

If you do this, The AudioSample c'tor should take the buffer and index parameters and be constructed in-place in the map (emplace) to avoid the d'tor being called on the temporary copy. Deleting the copy and assignment c'tors would also be a good idea.